packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/flat_dast/untagged_flat_dast.rs

```diff
@@ -0,0 +1,383 @@
+    /// This function recursively adds all children etc. of the `DastRoot`.
+    pub fn merge_dast_root(&mut self, dast: &DastRoot) {
+        for child in dast.children.iter() {
+            self.merge_content(child, None);
+        }
+    }
+
+    /// Update the the node at `idx` to be `node`. Any children of `node` are added to the `FlatRoot` as well,
+    /// but children of the old node are not removed. This function does no consistency checking, so it is
+    /// the _caller should be extra careful_.
+    pub fn update_content(
+        &mut self,
+        node: &DastElementContent,
+        idx: Index,
+        parent: Option<Index>,
+    ) -> UntaggedContent {
+        match node {
+            DastElementContent::Element(elm) => {
+                self.set_element(elm, idx, parent);
+                let ret = UntaggedContent::Ref(idx);
+                // Newly set elements have empty children and attributes. It is our
+                // job to fill them in.
+                let children = elm
+                    .children
+                    .iter()
+                    .map(|child| self.merge_content(child, Some(idx)))
+                    .collect();
+                let attributes: Vec<FlatAttribute> = elm
+                    .attributes
+                    .values()
+                    .map(|attr| {
+                        let children = attr
+                            .children
+                            .iter()
+                            .map(|child| match child {
+                                DastTextMacroContent::Text(txt) => {
+                                    DastElementContent::Text(txt.clone())
+                                }
+                                DastTextMacroContent::Macro(macro_) => {
+                                    DastElementContent::Macro(macro_.clone())
+                                }
+                                DastTextMacroContent::FunctionMacro(function_macro) => {
+                                    DastElementContent::FunctionMacro(function_macro.clone())
+                                }
+                            })
+                            .map(|child| self.merge_content(&child, Some(idx)))
+                            .collect();
+                        FlatAttribute {
+                            name: attr.name.clone(),
+                            parent: Some(idx),
+                            children,
+                            position: attr.position.clone(),
+                        }
+                    })
+                    .collect();
+                self.set_children(idx, children);
+                self.set_attributes(idx, attributes);
+                ret
+            }
+            DastElementContent::Error(err) => {
+                let ret = UntaggedContent::Ref(self.set_error(err, idx, parent));
+                if parent.is_none() {
+                    self.children.push(ret.clone());
```

Comment:
This line should lead to duplication of a reference to the node. There must have been an element at `idx` already, or `set_error` would panic. Which means, if everything were consistent beforehand (and the replaced node also had no parent, as should be the case), the root should already have a reference to the index of this node. With this push, it should now have two references (and this new one is at the end of the document.)

Reply:
Good catch!